### PR TITLE
ramips: add support for ipTIME A3004NS-dual

### PR DESCRIPTION
**Pull request by HexagonWin**
ipTIME A3004NS-dual is a 2.4/5GHz band router, based on Mediatek MT7621.

Specifications:
- SoC: MT7621 (880MHz)
- RAM: DDR3 256M
- Flash: SPI NOR 16MB
- WiFi:
 - 2.4GHz: MT7602E
 - 5GHz : MT7612E
- Ethernet:
 - 4x LAN
 - 1x WAN
- USB: 1 * USB3.0 port
- UART:
  - 3.3V, TX, RX, GND / 57600 8N1

Installation via web interface:
- 1. Flash Initramfs image using OEM Firmware's web GUI
- 2. Boot into OpenWrt and perform Sysupgrade with sysupgrade image.

Revert to stock firmware:
- 1. Boot into OpenWrt and perform Sysupgrade with OEM Stock Firmware image.

Thanks.

Signed-off-by: Yuchan Seo <hexagonwin@disroot.org>

### CODE REVIEW

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (mans0n) [inline target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:73]: The reg pair should be `<start length>`, not `<start end>`.

File: target/linux/ramips/image/mt7621.mk
Comment (mans0n) [inline target/linux/ramips/image/mt7621.mk:758]: Just curious, are you sure you need this? What happens if you omit this?

File: target/linux/ramips/image/mt7621.mk
Comment (mans0n) [inline target/linux/ramips/image/mt7621.mk:763]: Probably you'll also need `kmod-usb-ledtrig-usbport`.

**Review by mans0n — COMMENTED** [review]
Hi, a few comments below.

**Comment by mans0n** [comment]
> Note : The device is very similar to the ipTIME A8004T.

What made you think so? To me it seems it rather resembles WF2881.

File: target/linux/ramips/image/mt7621.mk
Comment (HexagonWin) [reply target/linux/ramips/image/mt7621.mk:758]: Hello, OpenWrt does not boot without the $(Device/uimage-lzma-loader).
According to a UART log the router gets lzma error and reboots. 
About $(Device/dsa-migration) : I think this was just copied from iptime-a8004t. I will try removing it and testing.

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:9]: Hello, for full name, should I then write "iptime,a3004ns-dual", "mediatek,mt7621-soc" then? Thanks.

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:10]: Thanks, I will edit this part.

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:73]: Thanks, then should I change it to 
`reg = <0x20000 0x10000` ? Thanks.

**Comment by HexagonWin** [comment]
> > Note : The device is very similar to the ipTIME A8004T.
> 
> What made you think so? To me it seems it rather resembles WF2881.

Oh, the patch was based on that device and I could get it to boot with only a tiny change.

File: target/linux/ramips/image/mt7621.mk
Comment (HexagonWin) [reply target/linux/ramips/image/mt7621.mk]: Hello, I added the package on the DEVICE_PACKAGES section. Thanks.

**Comment by HexagonWin** [comment]
Hello, I fixed those parts and pushed again. Thanks.

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (mans0n) [reply target/linux/ramips/dts/mt7621_iptime_a3004nd.dts]: Yes, and please change the other corresponding parts accordingly. (factory, firmware, ...)

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (mans0n) [reply target/linux/ramips/dts/mt7621_iptime_a3004nd.dts]: Yes, and please change the other corresponding parts accordingly. (factory, firmware, ...)

File: target/linux/ramips/image/mt7621.mk
Comment (mans0n) [reply target/linux/ramips/image/mt7621.mk:758]: Okay, then other iptime devices might need that remedy. Thanks for the info.
AFAIK you need to keep `$(Device/dsa-migration)` part.

**Comment by mans0n** [comment]
Please squash your changes into a single commit.
https://openwrt.org/docs/guide-developer/working-with-github-pr#squashing_commits

>>> Note : The device is very similar to the ipTIME A8004T.
>>
>> What made you think so? To me it seems it rather resembles WF2881.
>
> Oh, the patch was based on that device and I could get it to boot with only a tiny change.

That note implies the PCB of A3004NS-dual is almost same with that of A8004T, which is not the case. IMHO that line can be dropped.

File: target/linux/ramips/image/mt7621.mk
Comment (mans0n) [inline target/linux/ramips/image/mt7621.mk:756]: Please apply proper alphabetic sorting.

**Comment by HexagonWin** [comment]
Thanks. I modified based on your comments and pushed again.

**Comment by mans0n** [comment]
You still have two commits in this PR.

**Comment by HexagonWin** [comment]
Sorry! Fixed.

**Comment by mans0n** [comment]
The [From header](https://github.com/openwrt/openwrt/commit/068ae1c381e9067cf0e4dab1ac89a3d0757e944c.patch) should match the Signed-off-by tag, but your patch uses your github username. You can fix this with `git config -e` and `git commit --amend --no-edit --reset-author`.

And `image/Makefile` disappeared from your patch...

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (mans0n) [inline target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:162]: The device does not use jtag pins, it uses uart3 pins. So this line should be:
`groups = "wdt", "uart3", "i2c";`

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (mans0n) [inline target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:79]: `reg = <0x30000 0x10000>;`

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (mans0n) [inline target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:85]: `reg = <0x40000 0xfc0000>;`

**Comment by mans0n** [comment]
And the dts filename should be "mt7621_iptime_a3004ns-dual.dts".

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:162]: It also has a JTAG port in the board

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:79]: Thanks, I'll edit it that way.

File: target/linux/ramips/dts/mt7621_iptime_a3004nd.dts
Comment (mans0n) [reply target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:162]: Hmm, you're right. Anyway, if you want to use the pins for jtag purpose, it should not be here, as this node is for gpio pins.

**Comment by HexagonWin** [comment]
Hello, I fixed the issues. Thanks

**Comment by mans0n** [comment]
The author line still contains your github username. You can check it by `git show --format=fuller`.

File: target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts
Comment (mans0n) [inline target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:162]: "uart3" should be included here as the USB LED use the pin.

Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:162]: I'll add this, Thanks.

**Comment by HexagonWin** [comment]
> The author line still contains your github username. You can check it by `git show --format=fuller`.

For me it shows my real name, not alpha2x on both the "Comment" section and commits section. Where does it show the user name? Thanks.

**Comment by HexagonWin** [comment]
Fixed it

**Comment by mans0n** [comment]
Goto the commit page -> append `.patch` to URL -> Check "From" header in [this page](https://github.com/openwrt/openwrt/commit/1a8e1b7b00138554abe72db3fa405c6ed38c480e.patch)

**Comment by HexagonWin** [comment]
Ah.. got it. But how should I fix it?
But the Signed-off-by part shows my name, does the author line still matter..?
Thanks.

**Comment by mans0n** [comment]
That's the author line, and it should match SoB in the comment section. (Maybe that's why DCO bot keeps failing?)
`git config user.name "Full Name"`
`git commit --amend --no-edit --reset-author`

**Comment by HexagonWin** [comment]
You're right. The error seems to be gone now.
Fixed it.

**Review by mans0n — APPROVED** [review]
`Reviewed-by: Sungbo Eo <mans0n@gorani.run>`
Thanks for your quick feedback!
I'm just a random reviewer here, you'll need to wait for a core dev to get it merged.

File: target/linux/ramips/image/mt7621.mk
Comment (CHKDSK88) [inline target/linux/ramips/image/mt7621.mk:736]: This should be dropped. Your device wasn't supported before DSA migration.

Comment (HexagonWin) [reply target/linux/ramips/image/mt7621.mk:736]: Thanks, I'll remove it.

**Comment by HexagonWin** [comment]
Fixed it.

File: target/linux/ramips/image/mt7621.mk
Comment (CHKDSK88) [reply target/linux/ramips/image/mt7621.mk]: My bad, please revert it. Adrian tells me why here: https://github.com/openwrt/openwrt/pull/4416#discussion_r691544161

Comment (HexagonWin) [reply target/linux/ramips/image/mt7621.mk:736]: @CHKDSK88 Did it..

**Comment by HexagonWin** [comment]
@CHKDSK88 I tried building the firmware right in this state, and it is working properly.
Is there anything more that I should edit?
Thanks.

**Comment by CHKDSK88** [comment]
> @CHKDSK88 I tried building the firmware right in this state, and it is working properly.
> Is there anything more that I should edit?
> Thanks.

It looks fine for me.

**Comment by HexagonWin** [comment]
Hello, may I ask why this isn't being merged...??
I don't see anything more to edit..Thanks.

File: target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts
Comment (adschm) [inline target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:180]: remove empty line at EOF.

Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts]: @adschm Hello, I fixed it. Thanks.

File: target/linux/ramips/image/mt7621.mk
Comment (mans0n) [inline target/linux/ramips/image/mt7621.mk:742]: Please add `kmod-usb-ledtrig-usbport` here.

Comment (HexagonWin) [reply target/linux/ramips/image/mt7621.mk:742]: Odd, it disappeared.. Thanks.

Comment (HexagonWin) [reply target/linux/ramips/image/mt7621.mk:742]: I don't know what I did but the name just reverted.. I'll just fix them all and re-post. Sorry!

Comment (HexagonWin) [reply target/linux/ramips/image/mt7621.mk]: Done

**Comment by HexagonWin** [comment]
Hello, if there aren't more issues, can this please be merged? Thanks.

**Comment by HexagonWin** [comment]
@adschm I'm sorry but can you please check if there are more issues with the PR? Thanks.

File: target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts
Comment (adschm) [inline target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:162]: I don't see a GPIO for jtag?

Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:162]: Hello, I'm sorry but I don't know much about this, is another jtag node needed?
The jtag pins are available on board. The similar for the supported ipTIME A8004T, it's dts ( https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_iptime_a8004t.dts ) also seems to not have a jtag node. Thanks.

Comment (mans0n) [reply target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:162]: If you're going to use the pins for gpio purpose, then leave it as it is. (though I do not prefer it personally)
For jtag purpose, you might need to add a node like this:
```
	jtag {
		groups = "jtag";
		function = "jtag";
	};
```
...or skip it if the bootloader already enabled it.
AFAIK A8004T does not expose jtag pins so its setup is right.
I don't care much about it here, but I believe it's not a good practice to add a feature that has not been tested. Have you tested the gpio function with the jtag pins? Or did you add the node just because the pins were there?

Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:162]: @mans0n: I just left it since the pins were available.
Unfortunately I do not have any kinds of testing devices, especially none for jtag.

Comment (adschm) [reply target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:162]: I would have simply removed the jtag group?

Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:162]: Alright, I'll just remove it then.
Should I just remove the thing by changing
```
groups = "wdt", "jtag", "i2c", "uart3";
```
to?
```
groups = "wdt", "i2c", "uart3";
```

Comment (mans0n) [reply target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:162]: Precisely.

Comment (HexagonWin) [reply target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts]: Done.

**Comment by HexagonWin** [comment]
@adschm: I successfully checked the patch working on latest master again. Could you please check again if there are more issues with the PR? Thanks.

**Comment by mans0n** [comment]
Thanks! Merged as c33eb0372e7c.

**Pull request closed without merging**
